central/notifications/service/service_impl.go

```diff
@@ -0,0 +1,142 @@
+
+	ds datastore.DataStore
+}
+
+// RegisterServiceServer registers this service with the given gRPC Server.
+func (s *serviceImpl) RegisterServiceServer(server *grpc.Server) {
+	v1.RegisterNotificationServiceServer(server, s)
+}
+
+// RegisterServiceHandler registers this service with the given gRPC Gateway endpoint.
+func (s *serviceImpl) RegisterServiceHandler(ctx context.Context, mux *runtime.ServeMux, conn *grpc.ClientConn) error {
+	return v1.RegisterNotificationServiceHandler(ctx, mux, conn)
+}
+
+// AuthFuncOverride specifies the auth criteria for this API.
+func (s *serviceImpl) AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error) {
+	return ctx, authorizer.Authorized(ctx, fullMethodName)
+}
+
+// CountNotifications returns the number of notifications matching the request query.
+func (s *serviceImpl) CountNotifications(ctx context.Context, request *v1.CountNotificationsRequest) (*v1.CountNotificationsResponse, error) {
+	query := getQueryBuilderFromFilter(request.GetFilter()).ProtoQuery()
+	count, err := s.ds.CountNotifications(ctx, query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to count notifications")
+	}
+	return &v1.CountNotificationsResponse{Count: int64(count)}, nil
+}
+
+// GetNotification returns a specific notification based on its ID.
+func (s *serviceImpl) GetNotification(ctx context.Context, resource *v1.ResourceByID) (*v1.GetNotificationResponse, error) {
+	resourceID := resource.GetId()
+	notification, err := s.ds.GetNotificationByID(ctx, resourceID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get notification %q", resourceID)
+	}
+	return &v1.GetNotificationResponse{Notification: convertToServiceType(notification)}, err
+}
+
+// ListNotifications returns all notifications matching the request query.
+func (s *serviceImpl) ListNotifications(ctx context.Context, request *v1.ListNotificationsRequest) (*v1.ListNotificationsResponse, error) {
+	query := getQueryBuilderFromFilter(request.GetFilter()).ProtoQuery()
+	paginated.FillPagination(query, request.GetPagination(), maxPaginationLimit)
+	paginated.FillDefaultSortOption(
+		query,
+		&v1.QuerySortOption{
+			Field:    search.LastUpdatedTime.String(),
+			Reversed: true,
+		},
+	)
+
+	notifications, err := s.ds.ListNotifications(ctx, query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list notifications")
+	}
+	var respNotifications []*v1.Notification
+	for _, n := range notifications {
+		respNotifications = append(respNotifications, convertToServiceType(n))
+	}
+	return &v1.ListNotificationsResponse{Notifications: respNotifications}, nil
+}
+
+func getQueryBuilderFromFilter(filter *v1.NotificationsFilter) *search.QueryBuilder {
+	queryBuilder := search.NewQueryBuilder().
+		AddTimeRangeField(
+			search.CreatedTime,
+			protoconv.ConvertTimestampToTimeOrDefault(filter.GetFrom(), time.Unix(0, 0)),
+			protoconv.ConvertTimestampToTimeOrNow(filter.GetUntil()),
+		)
```

Comment:
Do I understand correctly that if `filter==nil` everything is returned because the range will be from epoch to now? While I guess this is what we want, I wonder if we should avoid any `WHERE` clauses in the query (any complex query actually) in such case.

Reply:
Yes correct. Since the date columns are indexed I doubt the `WHERE` clauses have significant impact even if they are unnecessary for `filter==nil`. But we can leave them out, I just didn't think of that.